flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClientTest.java

```diff
@@ -0,0 +1,352 @@
+        globalPool = new NetworkBufferPool(NUM_TOTAL_BUFFERS, NETWORK_BUFFER_SIZE);
+    }
+
+    @AfterEach
+    void after() {
+        globalPool.destroy();
+    }
+
+    @TestTemplate
+    void testWriteRecordsToEmptyStorageTiers() throws IOException {
+        int numSubpartitions = 10;
+        int numBuffersInPool = 10;
+        int bufferSize = 1024;
+        Random random = new Random();
+
+        TieredStorageMemoryManagerImpl storageMemoryManager =
+                createStorageMemoryManager(numBuffersInPool);
+        BufferAccumulator bufferAccumulator =
+                new HashBufferAccumulator(numSubpartitions, bufferSize, storageMemoryManager);
+
+        TieredStorageProducerClient tieredStorageProducerClient =
+                new TieredStorageProducerClient(
+                        numSubpartitions, false, bufferAccumulator, null, Collections.emptyList());
+
+        assertThatThrownBy(
+                        () ->
+                                tieredStorageProducerClient.write(
+                                        generateRandomData(bufferSize, random),
+                                        new TieredStorageSubpartitionId(0),
+                                        Buffer.DataType.DATA_BUFFER,
+                                        isBroadcast))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessageContaining("Failed to choose a storage tier");
+    }
+
+    @TestTemplate
+    void testEmptyMetricUpdater() throws IOException {
+        int numSubpartitions = 10;
+        int numBuffersInPool = 10;
+        int bufferSize = 1024;
+        Random random = new Random();
+
+        TieredStorageMemoryManagerImpl storageMemoryManager =
+                createStorageMemoryManager(numBuffersInPool);
+        BufferAccumulator bufferAccumulator =
+                new HashBufferAccumulator(numSubpartitions, bufferSize, storageMemoryManager);
+        TestingTierProducerAgent testingTierProducerAgent =
+                new TestingTierProducerAgent(false, numSubpartitions);
+
+        TieredStorageProducerClient tieredStorageProducerClient =
+                new TieredStorageProducerClient(
+                        numSubpartitions,
+                        false,
+                        bufferAccumulator,
+                        null,
+                        Collections.singletonList(testingTierProducerAgent));
+
+        assertThatThrownBy(
+                        () ->
+                                tieredStorageProducerClient.write(
+                                        generateRandomData(bufferSize, random),
+                                        new TieredStorageSubpartitionId(0),
+                                        Buffer.DataType.DATA_BUFFER,
+                                        isBroadcast))
+                .isInstanceOf(NullPointerException.class);
+    }
+
+    @TestTemplate
+    void testWriteRecordsToMultipleStorageTiers() throws IOException {
```

Comment:
This test is too complex, and the some test logic does not belong to `TieredStorageProducerClient`. What we need to test is the logic that only belongs `TieredStorageProducerClient`. Others like whether the buffer is written to the tier is not its concern.

Reply:
Simplify the logic by introducing a `TestingTierProducerAgent`.